Sources/AsyncHTTPClient/ConnectionPool/HTTPRequestStateMachine.swift

```diff
@@ -0,0 +1,467 @@
+        }
+    }
+
+    mutating func requestStreamFinished() -> Action {
+        switch self.state {
+        case .initialized:
+            preconditionFailure("Invalid state")
+        case .running(.streaming(let expectedBodyLength, let sentBodyBytes, _), let responseState):
+            if let expected = expectedBodyLength, expected != sentBodyBytes {
+                let error = HTTPClientError.bodyLengthMismatch
+
+                switch responseState {
+                case .initialized, .receivingBody:
+                    self.state = .failed(error)
+                case .endReceived:
+                    #warning("TODO: This needs to be fixed. @Cory: What does this mean here?")
+                    preconditionFailure("Unimplemented")
+                }
+
+                return .failRequest(error, closeStream: true)
+            }
+
+            self.state = .running(.endSent, responseState)
+            return .sendRequestEnd(startReadTimeoutTimer: self.idleReadTimeout)
+
+        case .running(.verifyRequest, _),
+             .running(.endSent, _):
+            preconditionFailure("Invalid state")
+
+        case .finished:
+            return .wait
+
+        case .failed:
+            return .wait
+        }
+    }
+
+    mutating func requestCancelled() -> Action {
+        switch self.state {
+        case .initialized, .running:
+            let error = HTTPClientError.cancelled
+            self.state = .failed(error)
+            return .failRequest(error, closeStream: true)
+        case .finished:
+            return .wait
+        case .failed:
+            return .wait
+        }
+    }
+
+    mutating func channelInactive() -> Action {
+        switch self.state {
+        case .initialized, .running:
+            let error = HTTPClientError.remoteConnectionClosed
+            self.state = .failed(error)
+            return .failRequest(error, closeStream: false)
+        case .finished:
+            return .wait
+        case .failed:
+            // don't overwrite error
+            return .wait
+        }
+    }
+
+    // MARK: - Response
+
+    mutating func receivedHTTPResponseHead(_ head: HTTPResponseHead) -> Action {
+        switch self.state {
+        case .initialized:
+            preconditionFailure("How can we receive a response head before sending a request head ourselves")
+
+        case .running(let requestState, .initialized):
+            switch requestState {
+            case .verifyRequest:
+                preconditionFailure("How can we receive a response head before sending a request head ourselves")
+            case .streaming, .endSent:
+                break
+            }
+
+            self.state = .running(requestState, .receivingBody(.waiting))
+            return .forwardResponseHead(head)
+
+        case .running(_, .receivingBody), .running(_, .endReceived), .finished:
+            preconditionFailure("How can we sucessfully finish the request, before having received a head")
```

Comment:
We can in principle receive multiple `head`s for 1XX responses, which this code does not tolerate. That's fine, but we should probably mark it here as an enhancement point for later.